### ISSUE TITLE
type inference with lambdas doesn't work (requires explicit types)

### ISSUE THREAD

**Issue by timotheecour** [nim-lang/Nim#7435]
in D this works:
```d
auto testCallFun(alias fun)() { return fun(0); }
void main(){
  import std.stdio:writeln;
  writeln(testCallFun!(x=>x*2)());
}
```
in nim apparently this doesn't unless explicit types are provided:
```nim
import future
proc testCallFun[T](fun: T): auto = result = fun(0)
proc main() =
  # not ok: Error: invalid type: 'GenericParam' in this context: 'proc (fun: proc (x: GenericParam): untyped){.noSideEffect, gcsafe, locks: <unknown>.}' for proc
  echo testCallFun(x => x * 2)

  #ok with explicit type:
  echo testCallFun((x:int) => x * 2)
```

This pattern is heavily used and quite useful for generic code.

* Is there any fundamental reason why this can't work in nim?

* if not, is that on the roadmap?

**Comment by Araq** [comment]
Not on the roadmap as I can't figure out how to infer this type myself. Use a template instead, maybe.

**Comment by andreaferretti** [comment]
I have problems to figure out the type as well! :-D

For context, Nim's type inference does not perform unification - such as Hindley-Milner languages - but propagates type information forward only.

**Comment by Araq** [comment]
> For context, Nim's type inference does not perform unification - such as Hindley-Milner languages - but propagates type information forward only.

Unification is forward only.

**Comment by Araq** [comment]
Yes, Haskell and ML use a more complex scheme which propagates back. But the term "unification" is perfectly fine with forward only.
